### PR TITLE
First attempt of a hypercube UI

### PR DESCRIPTION
**Pull request by gabbaxx**
Now hypercubes can be created using the big plus button in the lower right.
More columns can be added with the little plus button inside the hypercube cards

### CODE REVIEW

File: src/index.html [hunk 1]
```diff
@@ -3,6 +3,7 @@
 <meta charset="utf-8">
 <meta name="description" content="Gain Qlik Core data model insights quickly during its development and validation phases.">
 <meta name="viewport" content="width=device-width, initial-scale=1.0">
+<link href="//fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet">
```
Comment (hrigner) [inline src/index.html]: What icons are we using? We removed the material icons library earlier this week..

File: src/components/use/dimension-list.jsx [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import { useMemo } from 'react';
+import usePromise from 'react-use-promise';
+
+export default function useModel(app, def) {
```
Comment (hrigner) [inline src/components/use/dimension-list.jsx]: useModel -> useDimensionList

File: src/components/hypercube-table.jsx [hunk 3]
```diff
@@ -78,133 +88,129 @@ function noRowsRenderer() {
   return <div className="no-values">No values</div>;
 }
 
-const GLYPH_SIZE = 6;
-
+const GLYPH_SIZE = 8;
+const MAX_ROW_GLYTH_LENGTH = 40;
```
Comment (hrigner) [inline src/components/hypercube-table.jsx]: GLYTH -> GLYPH

File: src/components/cube.jsx [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+Cube.propTypes = {
+  app: PropTypes.object.isRequired,
+  tableData: PropTypes.object.isRequired,
+};
+
+// Cube.defaultProps = {
+//   lastReloadTime: null,
+// };
```
Comment (hrigner) [inline src/components/cube.jsx]: remove

File: src/components/cube-column-chooser.pcss [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+  }
+
+  .expression-list::-webkit-scrollbar-track {
+    background: white;
+  }
+
+  .expression-list::-webkit-scrollbar-thumb {
+    background: #d9d9d9;
+  }
+}
```
Comment (hrigner) [inline src/components/cube-column-chooser.pcss]: we do not want to mess with the scrollbars, try to use the browser scrollbars

**Comment by axelssonHakan** [comment]
Don know if it's a problem but should the sessionObject be "destroyed" then closing a "measure/field" table

**Comment by axelssonHakan** [comment]
Creating several tables will push the first outside the window (left side)

**Comment by hrigner** [comment]
The name of the hypercube is increasing with 1 , Hypercube 1 -> Hypercube 2 -> Hypercube 3 etc. Can we just have "Hypercube", or is there some point in keeping track of the names?

File: src/components/cubes.jsx [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  function openColumnChooser() {
+    addOpen.current = true;
+    forceUpdate();
+  }
+
+  const popup = addOpen.current ? <CubeColumnChooser arrowStyle="arrowRight" alignTo={addButtonRef.current} selectableColumns={selectableColumns} chooseColumn={column => addCube(column)} /> : null;
+  const cubeDivs = cubeList.map(cube => (
+    <div key={cube.id} className="card">
+      <div className="title">{`HYPERCUBE ${cube.id}`}</div>
+      <i role="button" tabIndex="-1" className="material-icons close" onClick={() => removeCube(cube.id)}>close</i>
```
Comment (peol) [inline src/components/cubes.jsx]: material-icons are not more, use `<span>✖</span>` for now.

File: src/components/cubes.jsx [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+      <Cube app={app} tableData={cube} />
+    </div>));
+  return (
+    <div>
+      <div className="cubes">
+        <div className="add-button" ref={addButtonRef}>
+          <i role="button" tabIndex="-1" className="material-icons add" onClick={() => openColumnChooser()}>add_circle</i>
```
Comment (peol) [inline src/components/cubes.jsx]: Same here, use a regular character for add. @snurrfint and @hrigner are looking into SVG icons we can use.

File: src/components/hypercube-table.jsx [hunk 8]
```diff
@@ -5,6 +5,8 @@ import Column from 'react-virtualized/dist/es/Table/Column';
 import VirtualTable from './virtual-table';
 
 import './hypercube-table.pcss';
```
Comment (peol) [inline src/components/hypercube-table.jsx]: CSS last.

File: src/components/hypercube-table.jsx [hunk 9]
```diff
@@ -78,133 +88,129 @@ function noRowsRenderer() {
...
+const MAX_ROW_GLYTH_LENGTH = 40;
 function getMeasureWidth(layout, measureIndex, measureName) {
-  const dataSize = layout.qHyperCube.qMeasureInfo.length > 0 ? layout.qHyperCube.qMeasureInfo[measureIndex].qApprMaxGlyphCount : 0;
-  const titleSize = measureName.length;
-  const size = Math.max(dataSize, titleSize);
-  return size * GLYPH_SIZE;
+  const dataSize = layout.qHyperCube.qMeasureInfo.length > measureIndex ? layout.qHyperCube.qMeasureInfo[measureIndex].qApprMaxGlyphCount : 0;
+  const titleSize = (measureName || '').length;
+  const size = Math.min(Math.max(dataSize, titleSize), MAX_ROW_GLYTH_LENGTH);
+  return size * GLYPH_SIZE + 8;
```
Comment (peol) [inline src/components/hypercube-table.jsx]: What is `+ 8`?

File: src/components/cubes.jsx [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  const [, forceUpdate] = useState(null);
+  const addButtonRef = useRef(null);
+
+  const selectableColumns = useColumnOptions(app);
+  const addOpen = useRef(false); // Start with add dialog open
+
+  function addCube(column) {
+    if (column) {
+      const newCube = { id: count, initialColumns: [column] };
+      setCount(count + 1);
```
Comment (peol) [inline src/components/cubes.jsx]: I can't really see why we'd need `count` as a separate state, use `cubeList.length`?

File: src/components/cubes.pcss [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    display: flex;
+    flex-direction: column;
+    margin: 1em;
+    padding: 1em;
+    pointer-events: all;
+    position: relative;
+
+    .title {
+      color: #3f9dd0;
+      font-size: 1.25rem;
```
Comment (peol) [inline src/components/cubes.pcss]: We're mixing `rem` and `em` A LOT now. Can be go through the code base later and choose one?

File: src/components/cube-column-chooser.jsx [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+  const filteredColumnsOptions = selectableColumns.filter(item => item.title.toLowerCase().indexOf(filter) >= 0);
+
+
+  function updateFilter(e) {
+    const { value } = e.target;
+    setFilter(value.toLowerCase());
+  }
+
+  function selectRow(e) {
+    const title = findAttribute(e, 'data-title');
```
Comment (peol) [inline src/components/cube-column-chooser.jsx]: Not too fond of us relying on the DOM to transport data, can we solve this differently?

File: src/components/hypercube-table.jsx [hunk 13]
```diff
@@ -78,133 +88,129 @@ function noRowsRenderer() {
...
-    super();
-    this.state = {};
-    this.selfRef = React.createRef();
+function getTotalWidth(layout, dimensions, measures) {
+  if (!layout) {
+    return 0;
   }
+  const dimSizes = dimensions.map((dim, dimensionIndex) => getDimensionWidth(layout, dimensionIndex, dim.title));
+  const mesSizes = measures.map((measure, measureIndex) => getMeasureWidth(layout, measureIndex, measure.title));
+  return 20 + dimSizes.reduce((a, b) => a + b, 0) + mesSizes.reduce((a, b) => a + b, 0);
```
Comment (peol) [inline src/components/hypercube-table.jsx]: More magic numbers.

File: src/components/hypercube-table.jsx [hunk 14]
```diff
@@ -78,133 +88,129 @@ function noRowsRenderer() {
...
-      await this.object.setProperties(hypercubeProps);
-    }
-    const layout = await this.object.getLayout();
-    this.setState({ layout, object: this.object });
+export default function HypercubeTable({
+  app, measures, dimensions, height, maxWidth, onHeaderClick,
+}) {
+  const hypercubeProps = {
+    qInfo: {
+      qId: 'measurebox1',
```
Comment (peol) [inline src/components/hypercube-table.jsx]: Are there more than one?

File: src/components/hypercube-table.jsx [hunk 15]
```diff
@@ -78,133 +88,129 @@ function noRowsRenderer() {
...
-      this.object = await app.createSessionObject(hypercubeProps);
-    } else {
-      await this.object.setProperties(hypercubeProps);
-    }
-    const layout = await this.object.getLayout();
-    this.setState({ layout, object: this.object });
+export default function HypercubeTable({
+  app, measures, dimensions, height, maxWidth, onHeaderClick,
+}) {
+  const hypercubeProps = {
```
Comment (peol) [inline src/components/hypercube-table.jsx]: Move this out to a separate function (like `createDefinition()`), easier to test in the future.

File: src/components/hypercube-table.jsx [hunk 16]
```diff
@@ -78,133 +88,129 @@ function noRowsRenderer() {
...
-            </VirtualTable>
-          </div>
-        </div>
-      );
-    }
-    return (<div />);
+  const model = useModel(app, hypercubeProps);
+  const layout = useLayout(model);
+
+  function onHeaderClickInternal(data) {
```
Comment (peol) [inline src/components/hypercube-table.jsx]: This shouldn't be needed.

File: src/components/use/column-options.js [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+
+export function layoutToList(layout) {
+  return layout;
+}
+
+export default function useColumnOptions(app) {
+  const model = useModel(app, genericObjectProperties);
+  const layout = useLayout(model);
+  if (layout) {
+    const measures = layout.qMeasureList.qItems ? layout.qMeasureList.qItems.map(measure => ({ title: measure.qData.title, type: 'measure', hyperCubeContent: { qLibraryId: measure.qInfo.qId } })) : [];
```
Comment (peol) [inline src/components/use/column-options.js]: You should probably use a state for these calculations (i.e. do them inside `useEffect()` conditionally).

File: src/components/use/dimension-list.jsx (hunk 2 above)
Comment (peol) [inline src/components/use/dimension-list.jsx]: Why can't this just use `useModel` with a dimension list definition? Shouldn't have to duplicate the `useModel` code.

**Comment by peol** [comment]
screen shot 2018-11-08 at 8 20 01 am

This icon position could be a bit better/smaller?

File: src/components/app.jsx [hunk 18]
```diff
@@ -8,9 +8,8 @@ import TopBar from './topbar';
 import Model from './model';
 import Splash from './splash';
 
-import Measurebox from './measurebox';
-
 import './app.pcss';
+import Cubes from './cubes';
```
[the remark this answers is not in the document]
Comment (hrigner) [reply src/components/app.jsx]: Hm, lint should catch this?

**Comment by axelssonHakan** [comment]
Huge thanks for the possibility to copy cell text :stuck_out_tongue_winking_eye:

File: src/components/app.jsx (hunk 18 above)
[the remark this answers is not in the document]
Comment (peol) [reply src/components/app.jsx]: Yes, I thought so too :)

**Pull request merged**